[RUNTIME] SURGE_THREADS=1 panics in async TCP server path

## Description

Running an async TCP server with `SURGE_THREADS=1` can panic with:

```text
panic: async_return outside poll
```

This was observed while investigating network latency in the external `surgekv` project:

https://github.com/vovakirdan/surgekv

The same server works with the default worker count and with `SURGE_THREADS=2` or `SURGE_THREADS=4` for the tested PING path. With `SURGE_THREADS=1`, the process prints the listen banner and then exits, causing clients to receive `connection refused`.

This should be treated as a separate runtime correctness issue from the network performance problems.

## Surge Version

```
surge 0.1.10-dev — "forge storms before they land"
commit: b6a2c18175dd
message: Merge pull request #101 from vovakirdan/hash-foundation-plan-main
built:  2026-06-16T11:20:45Z
```

Local repository HEAD while investigating:

```
7eaddbc2f140c1d43c6ad38288cc5114f5b97e2a
```

## Steps to Reproduce

Current repro using the external `surgekv` server:

```bash
git clone https://github.com/vovakirdan/surgekv.git
cd surgekv
surge build .
SURGE_THREADS=1 ./target/debug/surgekv --port 43595 --workers 8 --shards 8 --expiry-interval-ms 250
```

Then, from another terminal, try to run a small PING benchmark from the same `surgekv` checkout:

```bash
go run scripts/bench_load.go \
  -target surgekv \
  -host 127.0.0.1 \
  -port 43595 \
  -op ping \
  -clients 1 \
  -requests 60 \
  -keys 1 \
  -prepare=false \
  -timeout 2s
```

## Current Behavior

The server log shows:

```text
surgekv listening on 127.0.0.1:43595
panic: async_return outside poll
```

The client side sees connection failures because the server process has exited:

```text
errors=60 first_error=dial tcp 127.0.0.1:43595: connect: connection refused
target=surgekv op=ping clients=1 requests=60 keys=1 value_bytes=64 total_ms=0 rps=209668 avg_us=0 p50_us=0 p95_us=0 p99_us=0 errors=60
```

### Code Example

`surgekv` uses the async TCP path with these pieces:

```sg
let accept_res = net.accept(&listener).await();
...
let task = spawn serve_client_limited(...);
...
let read_res = net.read_some(&conn, read_cap).await();
let write_res = net.write_all(&conn, data).await();
```

Relevant repro files in the external `surgekv` repository:

- `server/serve.sg`
- `server/line.sg`
- `server/conn.sg`

Relevant runtime areas to inspect in this repository:

- `runtime/native/rt_async_state.c`
- `runtime/native/rt_async_task.c`
- `runtime/native/rt_async_poll.c`
- `runtime/native/rt_net.c`

### Error Output

```text
panic: async_return outside poll
```

## Expected Behavior

`SURGE_THREADS=1` should either:

1. run the async TCP server correctly on a single worker, or
2. fail early with a clear supported-mode diagnostic if native net async requires more than one runtime thread.

It should not panic with `async_return outside poll` after the server starts listening.

## Additional Context

The older `testdata/llvm_parity` net examples are not currently a clean minimal repro because some of them hit diagnostics around sending `@nosend` `TcpListener`/`TcpConn` through `spawn` and need `@local spawn` updates. This issue likely needs a reduced regression test in the main `surge` repo that uses the current sendability rules.

Suggested next step: extract a minimal async TCP accept/read/write server from the external `surgekv` repro into a runtime regression test, then run it with `SURGE_THREADS=1` and the default worker count.

--- a/internal/vm/mt_correctness_test.go
+++ b/internal/vm/mt_correctness_test.go
@@ -897,6 +897,243 @@ fn main(port: uint, count: uint) -> int {
 	}
 }
 
+func TestNativeNetSingleThreadBlockingChannelInAsyncServer(t *testing.T) {
+	ensureLLVMToolchain(t)
+
+	source := `import stdlib/net as net;
+
+fn pong() -> byte[] {
+    let mut out: byte[] = [];
+    out.push(80:byte);
+    out.push(79:byte);
+    out.push(78:byte);
+    out.push(71:byte);
+    out.push(10:byte);
+    return out;
+}
+
+tag Request(Channel<string>);
+type RequestMsg = Request(Channel<string>);
+
+fn manager_apply(requests: &Channel<RequestMsg>) -> string {
+    let reply = make_channel::<string>(1:uint);
+    let request: RequestMsg = Request(reply);
+    requests.send(own request);
+
+    return compare reply.recv() {
+        Some(response) => response;
+        nothing => "stopped";
+    };
+}
+
+async fn manager_run(requests: Channel<RequestMsg>) -> nothing {
+    let mut done: bool = false;
+    while !done {
+        compare requests.recv() {
+            Some(msg) => {
+                compare msg {
+                    Request(reply) => {
+                        let response: string = "OK";
+                        reply.send(own response);
+                    }
+                };
+            }
+            nothing => {
+                done = true;
+            }
+        };
+    }
+    return nothing;
+}
+
+async fn handle_client(handle: int, requests: Channel<RequestMsg>) -> nothing {
+    let conn: TcpConn = { __opaque: handle };
+    while true {
+        let read_res = net.read_some(&conn, 16:uint).await();
+        let read_ok: bool = compare read_res {
+            Success(net_read_res) => compare net_read_res {
+                Success(bytes) => bytes.__len() != 0:uint;
+                _ => false;
+            };
+            Cancelled() => false;
+        };
+        if !read_ok {
+            break;
+        }
+
+        let data = pong();
+        let write_res = net.write_all(&conn, data).await();
+        let write_ok: bool = compare write_res {
+            Success(net_write_res) => compare net_write_res {
+                Success(_) => true;
+                _ => false;
+            };
+            Cancelled() => false;
+        };
+        if !write_ok {
+            break;
+        }
+    }
+
+    let _ = manager_apply(&requests);
+    let _ = net.close_conn(own conn);
+    return nothing;
+}
+
+async fn serve_worker(clients: Channel<int>, requests: Channel<RequestMsg>) -> nothing {
+    let mut client_tasks: Task<nothing>[] = [];
+    let mut done: bool = false;
+    while !done {
+        compare clients.recv() {
+            Some(handle) => {
+                let client_requests = requests;
+                let client_task = spawn handle_client(handle, client_requests);
+                client_tasks.push(client_task);
+            }
+            nothing => {
+                done = true;
+            }
+        };
+    }
+
+    while client_tasks.__len() != 0:uint {
+        let client_task = client_tasks.pop().safe();
+        let _ = client_task.await();
+    }
+    return nothing;
+}
+
+async fn serve_one(listener: TcpListener) -> int {
+    let clients = make_channel::<int>(1:uint);
+    let requests = make_channel::<RequestMsg>(1:uint);
+    let manager_requests = requests;
+    let manager_task = spawn manager_run(manager_requests);
+    let worker_clients = clients;
+    let worker_requests = requests;
+    let worker_task = spawn serve_worker(worker_clients, worker_requests);
+
+    while true {
+        let accept_res = net.accept(&listener).await();
+        compare accept_res {
+            Success(conn_res) => {
+                compare conn_res {
+                    Success(conn) => {
+                        clients.send(conn.__opaque);
+                    }
+                    _ => {
+                        clients.close();
+                        let _ = worker_task.await();
+                        requests.close();
+                        let _ = manager_task.await();
+                        return 1;
+                    }
+                };
+            }
+            Cancelled() => {
+                clients.close();
+                let _ = worker_task.await();
+                requests.close();
+                let _ = manager_task.await();
+                return 1;
+            }
+        };
+    }
+
+    clients.close();
+    let _ = worker_task.await();
+    requests.close();
+    let _ = manager_task.await();
+    return 0;
+}
+
+@entrypoint("argv")
+fn main(port: uint) -> int {
+    let listen_res = net.listen("127.0.0.1", port);
+    compare listen_res {
+        Success(listener) => {
+            print("ready");
+            let task: Task<int> = @local spawn serve_one(listener);
+            return compare task.await() {
+                Success(code) => code;
+                Cancelled() => 99;
+            };
+        }
+        err => {
+            print("listen_err=" + (err.code to string));
+            return 1;
+        }
+    };
+    return 1;
+}
+`
+
+	outputPath := buildLLVMProgramFromSource(t, source)
+	port := pickFreePort(t)
+	cmd := exec.Command(outputPath, strconv.Itoa(port))
+	env := envWithStdlib(repoRoot(t))
+	env = overrideEnvVar(env, "SURGE_THREADS", "1")
+	env = overrideEnvVar(env, "SURGE_BLOCKING_THREADS", "1")
+	cmd.Env = env
+	var outBuf, errBuf bytes.Buffer
+	cmd.Stdout = &outBuf
+	cmd.Stderr = &errBuf
+
+	if err := cmd.Start(); err != nil {
+		t.Fatalf("start single-thread echo server: %v", err)
+	}
+
+	waitCh := make(chan error, 1)
+	go func() {
+		waitCh <- cmd.Wait()
+	}()
+	defer func() {
+		if cmd.Process != nil {
+			_ = cmd.Process.Kill()
+		}
+		select {
+		case <-waitCh:
+		case <-time.After(2 * time.Second):
+		}
+	}()
+
+	fail := func(format string, args ...any) {
+		if cmd.Process != nil {
+			_ = cmd.Process.Kill()
+		}
+		<-waitCh
+		t.Fatalf(format+"\nstdout:\n%s\nstderr:\n%s", append(args, outBuf.String(), errBuf.String())...)
+	}
+
+	addr := net.JoinHostPort("127.0.0.1", strconv.Itoa(port))
+	conn, err := dialWithRetry(addr, time.Now().Add(10*time.Second))
+	if err != nil {
+		fail("dial single-thread echo server: %v", err)
+	}
+	if err = conn.SetDeadline(time.Now().Add(5 * time.Second)); err != nil {
+		_ = conn.Close()
+		fail("set deadline: %v", err)
+	}
+	if _, err = conn.Write([]byte("PING\n")); err != nil {
+		_ = conn.Close()
+		fail("write ping: %v", err)
+	}
+	var buf [5]byte
+	if _, err = io.ReadFull(conn, buf[:]); err != nil {
+		_ = conn.Close()
+		fail("read pong: %v", err)
+	}
+	_ = conn.Close()
+	if string(buf[:]) != "PONG\n" {
+		fail("unexpected response: %q", string(buf[:]))
+	}
+
+	select {
+	case err := <-waitCh:
+		t.Fatalf("single-thread echo server exited early: %v\nstdout:\n%s\nstderr:\n%s", err, outBuf.String(), errBuf.String())
+	case <-time.After(2 * time.Second):
+	}
+}
+
 func runMTHTTPKeepaliveScenario(addr string) error {
 	conn, err := dialWithRetry(addr, time.Now().Add(10*time.Second))
 	if err != nil {

--- a/runtime/native/rt_async_internal.h
+++ b/runtime/native/rt_async_internal.h
@@ -275,7 +275,7 @@ __surge_poll_call(uint64_t id); // NOLINT(bugprone-reserved-identifier,cert-dcl3
 extern uint64_t __surge_blocking_call(uint64_t id, void* state);
 
 extern rt_executor exec_state;
-extern _Thread_local jmp_buf poll_env;
+extern _Thread_local jmp_buf* poll_env;
 extern _Thread_local int poll_active;
 extern _Thread_local poll_outcome poll_result;
 extern _Thread_local waker_key pending_key;

--- a/runtime/native/rt_async_poll.c
+++ b/runtime/native/rt_async_poll.c
@@ -47,27 +47,46 @@ static poll_outcome poll_sleep_task(const rt_executor* ex, rt_task* task) {
     return out;
 }
 
+static void restore_poll_context(jmp_buf* saved_env,
+                                 int saved_active,
+                                 poll_outcome saved_result,
+                                 waker_key saved_pending) {
+    poll_env = saved_env;
+    poll_active = saved_active;
+    poll_result = saved_result;
+    pending_key = saved_pending;
+}
+
 static poll_outcome poll_user_task(const rt_executor* ex, const rt_task* task) {
     poll_outcome out = {POLL_NONE, waker_none(), NULL, 0};
     if (ex == NULL || task == NULL) {
         out.kind = POLL_DONE_CANCELLED;
         return out;
     }
+    // Single-thread awaits can poll another task while an async poll is already active.
+    // Keep the outer poll terminator target intact for when the outer task resumes.
+    jmp_buf env;
+    jmp_buf* saved_env = poll_env;
+    int saved_active = poll_active;
+    poll_outcome saved_result = poll_result;
+    waker_key saved_pending = pending_key;
     pending_key = waker_none();
     poll_result.kind = POLL_NONE;
     poll_result.park_key = waker_none();
     poll_result.state = NULL;
     poll_result.value_bits = 0;
+    poll_env = &env;
     poll_active = 1;
-    if (setjmp(poll_env) == 0) {
+    if (setjmp(env) == 0) {
         __surge_poll_call((uint64_t)task->poll_fn_id);
-        poll_active = 0;
+        restore_poll_context(saved_env, saved_active, saved_result, saved_pending);
         panic_msg("async poll returned without terminator");
         out.kind = POLL_DONE_CANCELLED;
         return out;
     }
-    poll_active = 0;
-    return poll_result;
+    out = poll_result;
+    restore_poll_context(saved_env, saved_active, saved_result, saved_pending);
+    return out;
 }
 
 poll_outcome poll_task(rt_executor* ex, rt_task* task) {
@@ -246,7 +265,7 @@ void run_until_done(rt_executor* ex, const rt_task* task, uint8_t* out_kind, uin
 }
 
 void rt_async_yield(void* state) {
-    if (!poll_active) {
+    if (!poll_active || poll_env == NULL) {
         panic_msg("async_yield outside poll");
         return;
     }
@@ -256,7 +275,7 @@ void rt_async_yield(void* state) {
         poll_result.kind = POLL_DONE_CANCELLED;
         poll_result.park_key = waker_none();
         pending_key = waker_none();
-        longjmp(poll_env, 1);
+        longjmp(*poll_env, 1);
     }
     if (waker_valid(pending_key)) {
         poll_result.kind = POLL_PARKED;
@@ -266,11 +285,11 @@ void rt_async_yield(void* state) {
         poll_result.park_key = waker_none();
     }
     pending_key = waker_none();
-    longjmp(poll_env, 1);
+    longjmp(*poll_env, 1);
 }
 
 void rt_async_return(void* state, uint64_t bits) {
-    if (!poll_active) {
+    if (!poll_active || poll_env == NULL) {
         panic_msg("async_return outside poll");
         return;
     }
@@ -279,11 +298,11 @@ void rt_async_return(void* state, uint64_t bits) {
     poll_result.kind = POLL_DONE_SUCCESS;
     poll_result.park_key = waker_none();
     pending_key = waker_none();
-    longjmp(poll_env, 1);
+    longjmp(*poll_env, 1);
 }
 
 void rt_async_return_cancelled(void* state) {
-    if (!poll_active) {
+    if (!poll_active || poll_env == NULL) {
         panic_msg("async_cancel outside poll");
         return;
     }
@@ -292,5 +311,5 @@ void rt_async_return_cancelled(void* state) {
     poll_result.kind = POLL_DONE_CANCELLED;
     poll_result.park_key = waker_none();
     pending_key = waker_none();
-    longjmp(poll_env, 1);
+    longjmp(*poll_env, 1);
 }

--- a/runtime/native/rt_async_state.c
+++ b/runtime/native/rt_async_state.c
@@ -23,7 +23,7 @@
 // - virtual time still advances on yields; timers only fast-forward when the system is idle.
 
 rt_executor exec_state;
-_Thread_local jmp_buf poll_env;
+_Thread_local jmp_buf* poll_env;
 _Thread_local int poll_active = 0;
 _Thread_local poll_outcome poll_result;
 _Thread_local waker_key pending_key;

--- a/runtime/native/rt_async_task.c
+++ b/runtime/native/rt_async_task.c
@@ -161,7 +161,9 @@ void rt_task_await(void* task, uint8_t* out_kind, uint64_t* out_bits) {
         rt_unlock(ex);
         return;
     }
+    rt_task* current = rt_current_task();
     run_until_done(ex, target, out_kind, out_bits);
+    rt_set_current_task(current);
     rt_lock(ex);
     task_release(ex, target);
     rt_unlock(ex);

